alt_e2eshark/e2e_testing/backends.py

```diff
@@ -0,0 +1,65 @@
+# Copyright 2024 Advanced Micro Devices, Inc.
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+import abc
+from typing import TypeVar
+from e2e_testing.storage import TestTensors
+
+CompiledArtifact = TypeVar("CompiledArtifact")
+Invoker = TypeVar("Invoker")
+
+# This file should contain customizations for how to compile mlir from various entrypoints
+
+
+class BackendBase(abc.ABC):
+
+    @abc.abstractmethod
+    def compile(self, module) -> CompiledArtifact:
+        """specifies how to compile an MLIR Module"""
+
+    @abc.abstractmethod
+    def load(self, artifact: CompiledArtifact, func_name: str) -> Invoker:
+        """loads the function with name func_name from compiled artifact. This method should return a function callable from python."""
+
+
+from iree import compiler as ireec
+from iree import runtime as ireert
+
+
+class SimpleIREEBackend(BackendBase):
+    '''This backend uses iree to compile and run MLIR modules for a specified hal_target_backend'''
+    def __init__(self, *, hal_target_backend="llvm-cpu"):
+        self.hal_target_backend = hal_target_backend
```

Comment:
Either the class name should be fully descriptive (`IREEBackendCPULocalTask`) or this should allow configuring both the compiler target backend (e.g. `llvm-cpu`) _and_ the runtime HAL device (e.g. `local-task`). For GPU backends like Vulkan, CUDA, and ROCm, the `ireert.Config("local-task")` call will use a different device like `ireert.Config("vulkan")`.

For naming and the set of configurations, see here for reference: https://github.com/iree-org/iree/tree/main/build_tools/pkgci/external_test_suite